core/dbt/include/global_project/macros/materializations/incremental/incremental.sql

```diff
@@ -25,12 +27,34 @@
       {% set build_sql = create_table_as(False, target_relation, sql) %}
       {% do to_drop.append(backup_relation) %}
   {% else %}
-      {% set tmp_relation = make_temp_relation(target_relation) %}
       {% do run_query(create_table_as(True, tmp_relation, sql)) %}
+
+      {% set schema_changed = check_for_schema_changes(tmp_relation, target_relation) %}
+
+      {% if schema_changed %}
+      
+        {% if on_schema_change=='fail' %}
+          {{ 
+            exceptions.raise_compiler_error('The source and target schemas on this incremental model are out of sync!
+               Please re-run the incremental model with full_refresh set to True to update the target schema.
+               Alternatively, you can update the schema manually and re-run the process.') 
+          }}
+      
+        {# unless we ignore, run the sync operation per the config #}
+        {% elif on_schema_change != 'ignore' %}
+        
+          {% set schema_changes = sync_columns(tmp_relation, target_relation, on_schema_change) %}
+
+        {% endif %}
+
+      {% endif %}
```

Comment:
Could we wrap this in a macro? We'll need to add this logic to every `incremental` materialization (including Snowflake and BigQuery), and I think it will be DRYer to call a macro rather than copy-paste.

I do think it makes a ton of sense for each incremental materialization to be responsible for including logic like:
```sql
{% set on_schema_change = incremental_validate_on_schema_change(config.get('on_schema_change')) %}
{% do process_schema_changes(on_schema_change, tmp_relation, target_relation) %}
```
Where `process_schema_changes` is a macro that wraps logic for the `on_schema_change`-specific behavior, whether raising the compilation error or running `sync_columns`.

Reply:
@jtcohen6 up good call to abstract this.